### PR TITLE
feat: add cursor-based pagination to transfers endpoints

### PR DESCRIPTION
**Pull request by sborrazas**
This is a preamble to implementing the ability to filter transfers by
a txis scope.

### CODE REVIEW

**Review by thepiwo — APPROVED** [review]

File: test/integration/ae_mdw_web/controllers/transfer_controller_test.exs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+               String.starts_with?(kind, kind_prefix) and first_gen <= height and
+                 height <= last_gen
+             end)
+    end
+
+    test "when providing account and kind prefix filters, it returns transfers filtered by account and kind",
+         %{conn: conn} do
+      account_pk = "ak_21rna3xrD7p32U3vpXPSmanjsnSGnh6BWFPC9Pe7pYxeAW8PpS"
+
+      conn = get(conn, "/transfers/forward?account=#{account_pk}")
```
Comment (jyeshe) [inline test/integration/ae_mdw_web/controllers/transfer_controller_test.exs:170]: missing kind prefix

File: test/integration/ae_mdw_web/controllers/transfer_controller_test.exs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      heights2 = Enum.map(response_next["data"], fn %{"height" => height} -> height end)
+
+      assert Enum.all?(heights2, fn height -> last <= height and height <= first end)
+      assert Enum.at(heights2, 0) <= Enum.at(heights, limit - 1)
+    end
+
+    test "when providing account filter, it returns transfers filtered by account", %{conn: conn} do
+      account_pk = "ak_21rna3xrD7p32U3vpXPSmanjsnSGnh6BWFPC9Pe7pYxeAW8PpS"
+
+      conn = get(conn, "/transfers/forward?account=#{account_pk}")
```
Comment (jyeshe) [inline test/integration/ae_mdw_web/controllers/transfer_controller_test.exs:96]: could we have also for backward (with account and/or kind)?

File: lib/ae_mdw_web/controllers/tx_controller.ex [hunk 3]
```diff
@@ -38,27 +41,27 @@ defmodule AeMdwWeb.TxController do
...
-    case Txs.fetch_txs(direction, scope, query, cursor, limit) do
-      {:ok, txs, new_cursor} ->
-        path =
-          case params do
-            %{"scope_type" => scope_type, "range" => range} -> "/txs/#{scope_type}/#{range}"
-            _params -> "/txs/#{direction}"
-          end
+    %{direction: direction, limit: limit, cursor: cursor, scope: scope} = assigns
+
+    with {:ok, query} <- extract_query(query_params),
```
Comment (jyeshe) [inline lib/ae_mdw_web/controllers/tx_controller.ex:46]: thanks for specializing

File: lib/ae_mdw_web/controllers/tx_controller.ex [hunk 4]
```diff
@@ -38,27 +41,27 @@ defmodule AeMdwWeb.TxController do
...
 
-            URI.to_string(%URI{path: path, query: URI.encode_query(next_params)})
-          end
-
-        txs = handle_spendtx_details(txs, params)
+          URI.to_string(%URI{path: path, query: URI.encode_query(next_params)})
+        end
 
-        json(conn, %{"data" => txs, "next" => uri})
+      txs = handle_spendtx_details(txs, params)
```
Comment (jyeshe) [inline lib/ae_mdw_web/controllers/tx_controller.ex:61]: :+1:

File: lib/ae_mdw_web/controllers/tx_controller.ex [hunk 5]
```diff
@@ -165,6 +168,48 @@ defmodule AeMdwWeb.TxController do
     json(conn, maybe_add_spend_tx_details(Format.to_map(model_tx)))
   end
 
+  defp extract_query(query_params) do
+    query_params
+    |> Enum.reject(fn {key, _val} -> key in @pagination_param_keys end)
+    |> Enum.reduce_while({:ok, %{}}, fn {key, val}, {:ok, top_level} ->
```
Comment (jyeshe) [inline lib/ae_mdw_web/controllers/tx_controller.ex:174]: no big deal but the tagging of the result could be handled after the reduce_while

**Review by jyeshe — COMMENTED** [review]
Code and test look good (only one to be fixed). Some more tests for `backward` direction would be good for coverage.

**Comment by thepiwo** [comment]
let get this reviewed and merged to be in our next release :)

File: test/integration/ae_mdw_web/controllers/transfer_controller_test.exs (hunk 1 above)
Comment (sborrazas) [reply test/integration/ae_mdw_web/controllers/transfer_controller_test.exs]: Done

File: test/integration/ae_mdw_web/controllers/transfer_controller_test.exs (hunk 2 above)
Comment (sborrazas) [reply test/integration/ae_mdw_web/controllers/transfer_controller_test.exs:96]: Done

File: lib/ae_mdw_web/controllers/tx_controller.ex (hunk 5 above)
Comment (sborrazas) [reply lib/ae_mdw_web/controllers/tx_controller.ex:174]: I agree. I just moved this code from the plug to the controller, because it was tx-specific

**Review by jyeshe — APPROVED** [review]
test passes

**Pull request merged**